<a href="https://colab.research.google.com/github/Aishwarya-ps-kumbla/MST_Aish_practice_problems_01/blob/main/assignment07.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [59]:
#importing necessary libraries
import pandas as pd
import numpy as np


In [83]:
#reading the dataset
data=pd.read_csv('/content/tweets.csv',encoding="utf-8")

In [84]:
#checking the shape of the data
data.shape

(7920, 3)

In [85]:
#viewing the first 3 rows of dataset
data.head(3)

id  label  \
0   1      0   
1   2      0   
2   3      0   

                                                                                                                                 tweet  
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone  
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/  
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu

In [86]:
#choosing relevent columns from dataset
data=data[['tweet','label']]

In [87]:
#viewing maximum column width so can get maximum view of the tweet
pd.set_option('display.max_colwidth',None)

In [88]:
#viewing dataset
data.head(3)

tweet  \
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone   
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/   
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu   

   label  
0      0  
1      0  
2      0

In [89]:
#Data Preprocessing

#importing string library
import string

#function to remove punctuations
def remove_punctuation(text):
  punctuationfree=' '
  punctuationfree="".join([i for i in text if i not in string.punctuation])
  return punctuationfree

#adding punctuationfree column to the dataset
data['clean_tweet']=data['tweet'].apply(lambda x:remove_punctuation(x))


In [90]:
#viewing dataset
data.head()

tweet  \
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone   
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/   
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu   
3                     I'm wired I know I'm George I was made that way ;) #iphone #cute #daventry #home http://instagr.am/p/Li_5_ujS4k/   
4         What amazing service! Apple won't even talk to me about a question I have unless I pay them $19.95 for their stupid support!   

   label  \
0      0   
1      0   
2      0   
3      0   
4      1   

                                                                                                               clean_tweet  
0         fingerprint Pregnancy Test httpsgooglh1MfQV android apps beautiful cute health igers iphoneonly iphonesia iphone  
1      Finally a transparant silicon case  Thanks to my uncle  yay Sony Xperia S sonyexperias… httpinstagramcompYGEt5JC6JM  
2                We love this Would you go talk makememories unplug relax iphone smartphone wifi connect httpfbme6N3LsUpCu  
3                          Im wired I know Im George I was made that way  iphone cute daventry home httpinstagrampLi5ujS4k  
4  What amazing service Apple wont even talk to me about a question I have unless I pay them 1995 for their stupid support

In [91]:
#to see the punctuations in the string library
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [92]:
#word tokenization function

import nltk
nltk.download('punkt')
def tokenization(text):
  tokens=nltk.word_tokenize(text)
  return tokens
#adding tokenized column to dataset
data['tokenized_tweet']=data['clean_tweet'].apply(lambda x:tokenization(x))

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [93]:
#viewing dataset
data.head()

tweet  \
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone   
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/   
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu   
3                     I'm wired I know I'm George I was made that way ;) #iphone #cute #daventry #home http://instagr.am/p/Li_5_ujS4k/   
4         What amazing service! Apple won't even talk to me about a question I have unless I pay them $19.95 for their stupid support!   

   label  \
0      0   
1      0   
2      0   
3      0   
4      1   

                                                                                                               clean_tweet  \
0         fingerprint Pregnancy Test httpsgooglh1MfQV android apps beautiful cute health igers iphoneonly iphonesia iphone   
1      Finally a transparant silicon case  Thanks to my uncle  yay Sony Xperia S sonyexperias… httpinstagramcompYGEt5JC6JM   
2                We love this Would you go talk makememories unplug relax iphone smartphone wifi connect httpfbme6N3LsUpCu   
3                          Im wired I know Im George I was made that way  iphone cute daventry home httpinstagrampLi5ujS4k   
4  What amazing service Apple wont even talk to me about a question I have unless I pay them 1995 for their stupid support   

                                                                                                                                   tokenized_tweet  
0                   [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]  
1                [Finally, a, transparant, silicon, case, Thanks, to, my, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]  
2                        [We, love, this, Would, you, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]  
3                                  [Im, wired, I, know, Im, George, I, was, made, that, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]  
4  [What, amazing, service, Apple, wont, even, talk, to, me, about, a, question, I, have, unless, I, pay, them, 1995, for, their, stupid, support]

In [94]:
#function to remove stopwords
nltk.download('stopwords')
stopwords=nltk.corpus.stopwords.words('english')
def remove_stopwords(text):
  output= [i for i in text if i not in stopwords]
  return output
#adding stopwords removed column to dataset
data['remove_stopwords']=data['tokenized_tweet'].apply(lambda x:remove_stopwords(x))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [95]:
#viewing dataset
data.head()

tweet  \
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone   
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/   
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu   
3                     I'm wired I know I'm George I was made that way ;) #iphone #cute #daventry #home http://instagr.am/p/Li_5_ujS4k/   
4         What amazing service! Apple won't even talk to me about a question I have unless I pay them $19.95 for their stupid support!   

   label  \
0      0   
1      0   
2      0   
3      0   
4      1   

                                                                                                               clean_tweet  \
0         fingerprint Pregnancy Test httpsgooglh1MfQV android apps beautiful cute health igers iphoneonly iphonesia iphone   
1      Finally a transparant silicon case  Thanks to my uncle  yay Sony Xperia S sonyexperias… httpinstagramcompYGEt5JC6JM   
2                We love this Would you go talk makememories unplug relax iphone smartphone wifi connect httpfbme6N3LsUpCu   
3                          Im wired I know Im George I was made that way  iphone cute daventry home httpinstagrampLi5ujS4k   
4  What amazing service Apple wont even talk to me about a question I have unless I pay them 1995 for their stupid support   

                                                                                                                                   tokenized_tweet  \
0                   [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]   
1                [Finally, a, transparant, silicon, case, Thanks, to, my, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]   
2                        [We, love, this, Would, you, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]   
3                                  [Im, wired, I, know, Im, George, I, was, made, that, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]   
4  [What, amazing, service, Apple, wont, even, talk, to, me, about, a, question, I, have, unless, I, pay, them, 1995, for, their, stupid, support]   

                                                                                                                 remove_stopwords  
0  [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]  
1          [Finally, transparant, silicon, case, Thanks, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]  
2                  [We, love, Would, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]  
3                            [Im, wired, I, know, Im, George, I, made, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]  
4                           [What, amazing, service, Apple, wont, even, talk, question, I, unless, I, pay, 1995, stupid, support]

In [96]:
#function for lemmatization
from nltk.stem import WordNetLemmatizer # Capitalize 'W' in 'WordNetLemmatizer'
nltk.download('wordnet')
wordnet_lemmatizer=WordNetLemmatizer()
def lemmatizer(text):
  lemm_text = [wordnet_lemmatizer.lemmatize(word) for word in text]
  return lemm_text
#adding lemmatized column to dataset
data['lemmatized_tweet']=data['remove_stopwords'].apply(lambda x:lemmatizer(x))

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [97]:
#viewing dataset
data.head()

tweet  \
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone   
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/   
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu   
3                     I'm wired I know I'm George I was made that way ;) #iphone #cute #daventry #home http://instagr.am/p/Li_5_ujS4k/   
4         What amazing service! Apple won't even talk to me about a question I have unless I pay them $19.95 for their stupid support!   

   label  \
0      0   
1      0   
2      0   
3      0   
4      1   

                                                                                                               clean_tweet  \
0         fingerprint Pregnancy Test httpsgooglh1MfQV android apps beautiful cute health igers iphoneonly iphonesia iphone   
1      Finally a transparant silicon case  Thanks to my uncle  yay Sony Xperia S sonyexperias… httpinstagramcompYGEt5JC6JM   
2                We love this Would you go talk makememories unplug relax iphone smartphone wifi connect httpfbme6N3LsUpCu   
3                          Im wired I know Im George I was made that way  iphone cute daventry home httpinstagrampLi5ujS4k   
4  What amazing service Apple wont even talk to me about a question I have unless I pay them 1995 for their stupid support   

                                                                                                                                   tokenized_tweet  \
0                   [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]   
1                [Finally, a, transparant, silicon, case, Thanks, to, my, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]   
2                        [We, love, this, Would, you, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]   
3                                  [Im, wired, I, know, Im, George, I, was, made, that, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]   
4  [What, amazing, service, Apple, wont, even, talk, to, me, about, a, question, I, have, unless, I, pay, them, 1995, for, their, stupid, support]   

                                                                                                                 remove_stopwords  \
0  [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]   
1          [Finally, transparant, silicon, case, Thanks, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]   
2                  [We, love, Would, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]   
3                            [Im, wired, I, know, Im, George, I, made, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]   
4                           [What, amazing, service, Apple, wont, even, talk, question, I, unless, I, pay, 1995, stupid, support]   

                                                                                                                 lemmatized_tweet  
0  [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]  
1          [Finally, transparant, silicon, case, Thanks, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]  
2                  [We, love, Would, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]  
3                            [Im, wired, I, know, Im, George, I, made, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]  
4                           [What, amazing, service, Apple, wont, even, talk, question, I, unless, I, pay, 199

In [98]:
#assigning lemmatized tokens to corpus list
corpus = [' '.join(tokens) for tokens in data['lemmatized_tweet']]

In [99]:
#viewing first 10 sentences/strings in corpus
corpus[0:10]

['fingerprint Pregnancy Test httpsgooglh1MfQV android apps beautiful cute health igers iphoneonly iphonesia iphone',
 'Finally transparant silicon case Thanks uncle yay Sony Xperia S sonyexperias… httpinstagramcompYGEt5JC6JM',
 'We love Would go talk makememories unplug relax iphone smartphone wifi connect httpfbme6N3LsUpCu',
 'Im wired I know Im George I made way iphone cute daventry home httpinstagrampLi5ujS4k',
 'What amazing service Apple wont even talk question I unless I pay 1995 stupid support',
 'iPhone software update fucked phone big time Stupid iPhones',
 'Happy u instapic instadaily u sony xperia xperiaZ httpsinstagramcompz9qGfWlvj7',
 'New Type C charger cable UK httpwwwebaycoukitm112598674021 … bay Amazon etsy New Year Rob Cross Toby Young EVEMUN McMafia Taylor SPECTRE 2018 NewYear Starting 2018 recipe technology SamsungGalaxyS9 iPhoneX pictwittercomPjIwq59WtC',
 'Bout go shopping listening music iphone justme music likeforlike followforfollow… httpinstagrampVj6bg5tLql',


In [23]:
#Bag of words

In [100]:
#vectorization
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(ngram_range=(1,2))
#splitting data as x and y
x=cv.fit_transform(corpus)
y=data.label

In [101]:
#modelling using random forest classifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42)
clf=RandomForestClassifier()
clf.fit(x_train,y_train)

RandomForestClassifier()

In [102]:
#predicting the model and checking accuracy
from sklearn import metrics
y_pred=clf.predict(x_test)
print("Accuracy:",metrics.accuracy_score(y_pred,y_test))

Accuracy: 0.8383838383838383


In [103]:
#function to input sentiment and test using the model
def find_sentiment(text):
  input_text = cv.transform([text]) # Transform the input text, rename variable to avoid shadowing built-in 'input'
  pred = clf.predict(input_text)[0] # Get the prediction for the input
  if pred == 1:
    return "Negative"
  elif pred == 0:
    return "Positive"


In [104]:
#negative sentiment from corpus
input= 'What amazing service Apple wont even talk question I unless I pay 1995 stupid support'
find_sentiment(input)

'Negative'

In [105]:
#positive sentiment from corpus
input= 'Im wired I know Im George I was made that way iphone cute daventry home httpinstagrampLi5ujS4k'
find_sentiment(input)

'Positive'

In [30]:
#TFIDF
from sklearn.feature_extraction.text import TfidfVectorizer


In [107]:
#vectorization
def vectorizer_tfidf(corpus):
  tfidf_model=TfidfVectorizer()
  tfidf_matrix=tfidf_model.fit_transform(corpus).todense()
  tfidf_df=pd.DataFrame(tfidf_matrix)
  tfidf_df.columns=sorted(tfidf_model.vocabulary_)
  return tfidf_df

In [108]:
df1=vectorizer_tfidf(corpus)

In [109]:
df1

000  00000  002  004  0051  007  008   01  010111  0101am  ...  서울시장님  \
0     0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
1     0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
2     0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
3     0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
4     0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
...   ...    ...  ...  ...   ...  ...  ...  ...     ...     ...  ...    ...   
7915  0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
7916  0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
7917  0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
7918  0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   
7919  0.0    0.0  0.0  0.0   0.0  0.0  0.0  0.0     0.0     0.0  ...    0.0   

      안녕하세요  제야의종   종각  타종행사  𝒸𝓇𝑒𝒶𝓉𝓊𝓇𝑒  𝒹𝒶𝓃𝑔𝑒𝓇𝓊𝓈   𝒾𝓈  𝓌𝑒𝓁𝓁𝓇𝑒𝒶𝒹  𝓌𝓂𝒶𝓃  
0       0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
1       0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
2       0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
3       0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
4       0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
...     ...   ...  ...   ...       ...       ...  ...       ...   ...  
7915    0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
7916    0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
7917    0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
7918    0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  
7919    0.0   0.0  0.0   0.0       0.0       0.0  0.0       0.0   0.0  

[7920 rows x 22124 columns]

In [110]:
#modelling
tfidf=TfidfVectorizer(ngram_range=(1,2))
x=tfidf.fit_transform(corpus)
y=data.label

In [111]:
#fitting the model
clft=RandomForestClassifier(n_estimators=100)
clft.fit(x_train,y_train)

RandomForestClassifier()

In [112]:
#prediction and accuracy
y_pred=clft.predict(x_test)
print("Accuracy:",metrics.accuracy_score(y_pred,y_test))

Accuracy: 0.8354377104377104


In [113]:
#function to input sentiment
def analyze_sentiment(text):
  input_text = cv.transform([text]) # Transform the input text, rename variable to avoid shadowing built-in 'input'
  pred = clft.predict(input_text)[0] # Get the prediction for the input
  if pred == 1:
    return "Negative"
  elif pred == 0:
    return "Positive"

In [38]:
input= 'What amazing service Apple wont even talk question I unless I pay 1995 stupid support'
analyze_sentiment(input)

'Negative'

In [39]:
input= 'Im wired I know Im George I was made that way iphone cute daventry home httpinstagrampLi5ujS4k'
analyze_sentiment(input)

'Positive'

In [75]:
#Word2vec
from gensim.models import Word2Vec
from nltk import word_tokenize

In [114]:
#creating the model
model=Word2Vec(corpus,min_count=1,vector_size=100)

In [115]:
model

In [43]:
words=list(model.wv.index_to_key)

In [116]:
import gensim
#preprocessing
data['clean_data']=data['tweet'].apply(lambda x: gensim.utils.simple_preprocess(x))

In [117]:
#viewing the clean data after preprocessing
data.head()

tweet  \
0     #fingerprint #Pregnancy Test https://goo.gl/h1MfQV #android #apps #beautiful #cute #health #igers #iphoneonly #iphonesia #iphone   
1  Finally a transparant silicon case ^^ Thanks to my uncle :) #yay #Sony #Xperia #S #sonyexperias… http://instagram.com/p/YGEt5JC6JM/   
2          We love this! Would you go? #talk #makememories #unplug #relax #iphone #smartphone #wifi #connect... http://fb.me/6N3LsUpCu   
3                     I'm wired I know I'm George I was made that way ;) #iphone #cute #daventry #home http://instagr.am/p/Li_5_ujS4k/   
4         What amazing service! Apple won't even talk to me about a question I have unless I pay them $19.95 for their stupid support!   

   label  \
0      0   
1      0   
2      0   
3      0   
4      1   

                                                                                                               clean_tweet  \
0         fingerprint Pregnancy Test httpsgooglh1MfQV android apps beautiful cute health igers iphoneonly iphonesia iphone   
1      Finally a transparant silicon case  Thanks to my uncle  yay Sony Xperia S sonyexperias… httpinstagramcompYGEt5JC6JM   
2                We love this Would you go talk makememories unplug relax iphone smartphone wifi connect httpfbme6N3LsUpCu   
3                          Im wired I know Im George I was made that way  iphone cute daventry home httpinstagrampLi5ujS4k   
4  What amazing service Apple wont even talk to me about a question I have unless I pay them 1995 for their stupid support   

                                                                                                                                   tokenized_tweet  \
0                   [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]   
1                [Finally, a, transparant, silicon, case, Thanks, to, my, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]   
2                        [We, love, this, Would, you, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]   
3                                  [Im, wired, I, know, Im, George, I, was, made, that, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]   
4  [What, amazing, service, Apple, wont, even, talk, to, me, about, a, question, I, have, unless, I, pay, them, 1995, for, their, stupid, support]   

                                                                                                                 remove_stopwords  \
0  [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]   
1          [Finally, transparant, silicon, case, Thanks, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]   
2                  [We, love, Would, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]   
3                            [Im, wired, I, know, Im, George, I, made, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]   
4                           [What, amazing, service, Apple, wont, even, talk, question, I, unless, I, pay, 1995, stupid, support]   

                                                                                                                 lemmatized_tweet  \
0  [fingerprint, Pregnancy, Test, httpsgooglh1MfQV, android, apps, beautiful, cute, health, igers, iphoneonly, iphonesia, iphone]   
1          [Finally, transparant, silicon, case, Thanks, uncle, yay, Sony, Xperia, S, sonyexperias…, httpinstagramcompYGEt5JC6JM]   
2                  [We, love, Would, go, talk, makememories, unplug, relax, iphone, smartphone, wifi, connect, httpfbme6N3LsUpCu]   
3                            [Im, wired, I, know, Im, George, I, made, way, iphone, cute, daventry, home, httpinstagrampLi5ujS4k]   
4                           [What, amazing, service, Apple, wont, even, talk, question, I, unless, I, pay

In [118]:
#training the model
x_train, x_test, y_train, y_test = train_test_split(data['clean_data'], data['label'], test_size=0.2, random_state=42)

In [119]:
w2v_model = gensim.models.Word2Vec(x_train,min_count=1)

In [121]:
#checking one word embedding for 'iphone'
w2v_model.wv['iphone']

array([-0.4283257 ,  0.899957  ,  0.97352064,  1.036     ,  0.21464057,
       -1.7943413 ,  0.2241961 ,  2.2699614 , -1.2218941 , -0.5964464 ,
       -0.34226826, -2.0665004 , -0.21887104,  0.76882935,  0.634921  ,
       -0.9352118 ,  0.5081453 , -0.62136275, -0.60859996, -1.600151  ,
        0.10264725,  0.2719663 ,  0.87698877, -1.1321753 ,  0.01974691,
        0.12362111, -1.2858063 , -0.31922114, -0.6618931 ,  0.0164987 ,
        0.9083133 , -0.22972551,  0.59525806, -1.5808163 , -1.0095226 ,
        1.5543655 ,  0.67142177, -0.8531112 , -0.45671675, -1.8431501 ,
        0.5855216 , -1.1289654 , -0.74498063,  0.34764448,  0.98042524,
        0.0293263 , -0.9776413 , -0.21891183,  1.3613404 ,  0.75730383,
        0.8590328 , -1.1317269 , -0.10655124, -0.10677017, -0.66016513,
        0.42698717,  0.47870943,  0.09790199, -0.70032394,  0.46002135,
        0.36999154,  0.5430959 ,  0.5986412 , -0.1492278 , -0.8854051 ,
        1.4486734 ,  0.57048696,  1.2099096 , -1.5952199 ,  1.72

In [122]:
#viewing words similar to 'iphone'
w2v_model.wv.most_similar('iphone')

[('pet', 0.994575023651123),
 ('bestoftheday', 0.9929071068763733),
 ('photo', 0.9920734167098999),
 ('april', 0.9918140172958374),
 ('jj', 0.9910785555839539),
 ('beauty', 0.9909728169441223),
 ('hot', 0.9908524751663208),
 ('technology', 0.990466296672821),
 ('nature', 0.9903959035873413),
 ('beautiful', 0.9902528524398804)]

In [123]:
words=w2v_model.wv.index_to_key

In [124]:
len(words)

19072

In [125]:
#initializing variables for vectors
x_train_vec=[]
x_test_vec=[]

In [126]:
#iterate through each word in x_train
for sentence in x_train:
  #initialize an empty list to store vectors for current list
  vector=[]
  for word in sentence:
    #check word in our vocabulry
    if word in words:
      vector.append(w2v_model.wv[word])
  x_train_vec.append(vector)

In [127]:
#iterate through each word in x_test
for sentence in x_test:
  #initialize an empty list to store vectors for current list
  vector=[]
  for word in sentence:
    #check word in our vocabulry
    if word in words:
      vector.append(w2v_model.wv[word])
  x_test_vec.append(vector)

In [128]:
# Initialize empty lists to store average word vectors
x_train_vect_avg = []
x_test_vect_avg = []

#Define a function to calculate average word vectors
def calc_avg_vect(v):
  if v: # Check if the list is not empty
    return np.mean(np.array(v), axis=0) # Convert list of vectors to NumPy array and calculate mean
  else:
    return np.zeros(100,dtype=float)

#Calculate average vectors for X_train_vec
for v in x_train_vec:
  x_train_vect_avg.append(calc_avg_vect(v))

#Calculate average vectors for X_test_vec
for v in x_test_vec:
  x_test_vect_avg.append(calc_avg_vect(v))

In [129]:
#modelling
clw=RandomForestClassifier(n_estimators=100,random_state=42)
clw.fit(x_train_vect_avg,y_train)

RandomForestClassifier(random_state=42)

In [134]:
#prediction and accuracy
from sklearn import metrics
y_pred=clw.predict(x_test_vect_avg)
print('Accuracy: ',metrics.accuracy_score(y_test,y_pred))

Accuracy:  0.8693181818181818


In [80]:
def test_sentiment(text):
  # Transform the input text using the same method as during training
  input_text_vect = [w2v_model.wv[word] for word in word_tokenize(text) if word in w2v_model.wv] # Access word vectors using []
  input_text_avg = calc_avg_vect(input_text_vect)  # Calculate the average word vector
  input_text_avg_reshaped = input_text_avg.reshape(1, -1)  # Reshape to match model input

  pred = clw.predict(input_text_avg_reshaped)[0]  # Get the prediction
  if pred == 1:
    return "Negative"
  elif pred == 0:
    return "Positive"

In [81]:
input= 'What amazing service Apple wont even talk question I unless I pay 1995 stupid support'
test_sentiment(input)

'Negative'

In [82]:
input= 'Im wired I know Im George I was made that way iphone cute daventry home httpinstagrampLi5ujS4k'
test_sentiment(input)

'Positive'

Bag of Words Accuracy: 0.8383838383838383
TFIDF Accuracy: 0.835016835016835
Word2Vec Accuracy: 0.8693181818181818